datumaro/plugins/splitter.py

```diff
@@ -0,0 +1,480 @@
+    DEFAULT_SUBSET_NAME)
+
+class TaskSpecificSplitter(Transform):
+    def __init__(self, dataset, seed):
+        super().__init__(dataset)
+
+        np.random.seed(seed)
+
+        self._subsets = {"train", "val", "test"} # output subset names
+        self._parts = []
+        self._length = 'parent'
+
+    def _validate_splits(self, splits, validset=None):
+        subsets = []
+        ratios = []
+        if validset == None:
+            validset = ["train", "val", "test"]
+        for subset, ratio in splits:
+            assert subset in validset, "Subset name \
+                must be one of {}, but got {}.".format(validset, subset)
+            assert ratio >= 0.0 and ratio <= 1.0, "Ratio is expected to be \
+                in the range [0, 1], but got %s for %s" % (ratio, subset)
+            subsets.append(subset)
+            ratios.append(float(ratio))
+        ratios = np.array(ratios)
+
+        total_ratio = np.sum(ratios)
+        if not abs(total_ratio - 1.0) <= 1e-7:
+            raise Exception(
+                "Sum of ratios is expected to be 1, got %s, which is %s" %
+                (splits, total_ratio))
+        return subsets, ratios
+
+    def _get_required(self, ratio):
+        min_value = np.max(ratio)
+        for i in ratio:
+            if i < min_value and i > 1e-7:
+                min_value = i
+        required = int (np.around(1.0) / min_value)
+        return required
+
+    def _get_sections(self, dataset_size, ratio):
+        n_splits = [int(np.around(dataset_size * r)) for r in ratio[:-1]]
+        n_splits.append(dataset_size - np.sum(n_splits))
+
+        # if there are splits with zero samples even if ratio is not 0,
+        # borrow one from the split who has one or more.
+        for ii, num_split in enumerate(n_splits):
+            if num_split == 0 and ratio[ii] > 1e-7:
+                midx = np.argmax(n_splits)
+                if n_splits[midx] > 0:
+                    n_splits[ii] += 1
+                    n_splits[midx] -= 1
+        sections = np.add.accumulate(n_splits[:-1])
+        return sections
+
+    def _group_by_attributes(self, items):
+        '''
+        Args:
+            items: list of (idx, ann). ann is the annotation from Label object.
+        Returns:
+            by_attributes: dict of { combination-of-attributes : list of index }
+        '''
+        ## group by attributes
+        by_attributes = dict()
+        for idx, ann in items:
+            attributes = tuple(sorted(ann.attributes.items()))
+            if attributes not in by_attributes:
+                by_attributes[attributes] = []
+            by_attributes[attributes].append(idx)
+        return by_attributes
+
+    def _split_indice(self, indice, group_name, ratio, required):
+        filtered_size = len(indice)
+        if required > filtered_size:
+            log.warning("Not enough samples for filtered group, '{}'".format(
```

Comment:
Please follow the `%` formatting style, used everywhere else in the project. You can also omit formatting at all, because `logging` can substitute values itself.

Reply:
Thanks for the comment. I'll replace them all.